### ISSUE TITLE
Add file-based credentials support for on-prem gRPC clients to talk TLS with xDS Server

### ISSUE THREAD

**Issue by gu0keno0** [grpc/grpc#32977]
### Is your feature request related to a problem? Please describe.
We are evaluating gRPC proxyless service mesh, and we have our own xDS server implementation that uses standard file-based m-TLS for Authn.

One potential issue is that c-core only allows 3 default credentials when talking with xDS Server: Fake, Insecure, and GoogleDefault  (https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/channel_creds_registry_init.cc#L71). Although c-core, Java, Go all provide ways to implement additional ChannelCredsFactory, libraries that use C-binding such as Python doesn't seem to support the API to register additional ChannelCredsFactory. So it seems that only Insecure and GoogleDefault would be supported in these libraries, and therefore, we cannot use the TLS certs in our on-prem datacenter for them.

### Describe the solution you'd like
Add a new ChannelCredsFactory in c-core so that file-based TLS certs will be supported by default in c-core, and Python / other language bindings that use c-core can directly use this factory as well by configuring it in the bootstrap.json file.

### Describe alternatives you've considered
We've also considered extending the Cython library to add the hook for registering additional ChannelCredsFactory, yet it would be nicer if c-core can directly support it by default.

### Additional context
N/A.

**Comment by markdroth** [comment]
This is something we've been meaning to do for a while but hadn't had the chance to do yet.

Do you need only basic TLS, or do you need the ability to set options to configure things like mTLS?

@ejona86 @dfawley

**Comment by ejona86** [comment]
> we have our own xDS server implementation that uses standard file-based m-TLS for Authn.

Looks like it is mTLS. And with client certificates we almost certainly would need the file-polling stuff for picking up new versions.

**Comment by gu0keno0** [comment]
@markdroth @ejona86 thanks for replying! To confirm, we indeed prefer mTLS in our fleet, and therefore it will be great to do mTLS between xDS client and xDS server as well.

Cert rotation is also really nice to have. Although I'm not sure how cert rotation works under gRPC streaming. Would it require xDS client to re-establish the ads stream when we rotate cert? It's probably just an implementation detail, yet I'm curious.

**Comment by ejona86** [comment]
Connections and cert rotation are fully independent. The cert rotation would be noticed by polling and the new cert would be loaded to be used with new connections. You can use MAX_CONNECTION_AGE from [gRFC A9](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md#proposal) to trigger a new connection for new RPCs. For an existing ADS stream, you either need to close it occasionally (my recommendation) or use MAX_CONNECTION_AGE_GRACE to close the connection and kill the stream.

**Comment by gu0keno0** [comment]
Agreed, if gRPC is going to add options to set MAX_CONNECTION_AGE / MAX_CONNECTION_AGE_GRACE for ADS connection, we would love to enable it. My general understanding is that the ADS client is internal in gRPC implementation, and currently the main interface to configure it is the Bootstrap file? Please correct if I'm missing anything.

Alternatively, I considered to just force the client to reconnect by doing a rolling restart of our xDS server fleet, yet I'm on the same page that letting clients recreate the ADS connections feels nicer and more flexible.

**Comment by ejona86** [comment]
> Agreed, if gRPC is going to add options to set MAX_CONNECTION_AGE / MAX_CONNECTION_AGE_GRACE for ADS connection

Those are server-side settings, and already available.

**Comment by gu0keno0** [comment]
Cool. Thanks for the point. Then it's very straightforward for us to set on our in-house xDS server implementation. Given that, it feels like cert rotation should be straightforward? E.g. the ads stream is terminated --> client attempts to recreate the stream --> new certs will be loaded naturally.

**Comment by markdroth** [comment]
Just to be clear about how this works, new certs take effect when establishing a new *connection*, not a new stream.  But if your server closes the existing connection (and therefore the stream on that connection), a new connection will be created when attempting to create a new stream.

But in any case, yes, cert reloading should not be a problem here.  The only missing piece is the ability to configure mTLS in the bootstrap file.

We'll discuss this later this week and put together a plan for doing this in a cross-language way.

**Comment by gu0keno0** [comment]
Thank you, @markdroth @ejona86 . This will greatly unblock our xDS adoption.

**Comment by markdroth** [comment]
Just a quick update: Our meeting got cancelled last week, so we'll plan to discuss this later this week instead.

**Comment by markdroth** [comment]
Here's a gRFC proposing a way to configure this: https://github.com/grpc/proposal/pull/372

I've got a work-in-progress implementation for C-core in #33234.
